### ISSUE TITLE
Merge Pipeline crashed problem.

### ISSUE THREAD

**Issue by geek-frio** [quickwit-oss/quickwit#3627]
**Describe the bug**
We encountered the following exception error in our production environment:

```
quickwit_actors::actor_context: exit activating-kill-switch actor=MergePublisher-dry-SQHO exit_status=Failure(Failed to publish splits.

Caused by:
    Splits `["01H569VFSJZ9D7FT2NPV7J6JYZ"]` are not in a deletable state.)
```

After analyzing the logs, I found that PlanMerge operation was triggered twice, which ultimately caused the failure of the execution of the publish_split method for the second time. This is because one of the splits in the replaced_split has already been marked as MarkedDelete, which doesn't match the expected Published state. For example, this split（01H569VFSJZ9D7FT2NPV7J6JYZ） appeared in both PlanMerge occurrences.

```
quickwit_indexing::actors::merge_planner: Planned merge operation. merge_operation=Merge(operation_type=Merge, merged_split_id=01H569XDG31W7T691DTJKHW7WT,splits=[01H569WD97J2R1W33K5PEBH673,01H569W3M24ZPP18F1H5KVG43A,01H569VFSJZ9D7FT2NPV7J6JYZ,]

quickwit_indexing::actors::merge_planner: Planned merge operation. merge_operation=Merge(operation_type=Merge, merged_split_id=01H569XD3FFVZAN807JSYJ710S,splits=[01H569VFSJZ9D7FT2NPV7J6JYZ,01H569V5P2PJH4CXBJKA84JWKJ,01H569VJCD3SCJP1RJV3D6P6Z6,])
```

This ultimately led to the failure of the subsequent publish_splits operation. After analyzing the code, I found that the triggering of merge operations occurred in two actions: one is the automatically scheduled PlanMerge, and the other is the triggering after the creation of NewSplit. Is it possible that the same split was selected twice for PlanMerge, resulting in this issue?

My quickwit version is v0.6.2.

**Comment by geek-frio** [comment]
The check failed code is here:
```rust
const PUBLISH_SPLITS_QUERY: &str = r#"
            -- Select the splits to update, regardless of their state.
            -- The left join make it possible to identify the splits that do not exist.
            WITH input_splits AS (
                SELECT input_splits.split_id, input_splits.expected_split_state, splits.actual_split_state
                FROM (
                    SELECT split_id, 'Staged' AS expected_split_state
                    FROM UNNEST($3) AS staged_splits(split_id)
                    UNION
                    SELECT split_id, 'Published' AS expected_split_state
                    FROM UNNEST($4) AS published_splits(split_id)
                ) input_splits
                LEFT JOIN (
                    SELECT split_id, split_state AS actual_split_state
                    FROM splits
                    WHERE
                        index_uid = $1
                        AND (split_id = ANY($3) OR split_id = ANY($4))
                    FOR UPDATE
                    ) AS splits
                USING (split_id)
            ),
            -- Update the index metadata with the new checkpoint.
            updated_index_metadata AS (
                UPDATE indexes
                SET
                    index_metadata_json = $2
                WHERE
                    index_uid = $1
                    AND NOT EXISTS (
                        SELECT 1
                        FROM input_splits
                        WHERE
                            actual_split_state != expected_split_state
                        )
            ),
            -- Publish the staged splits and mark the published splits for deletion.
            updated_splits AS (
                UPDATE splits
                SET
                    split_state = CASE split_state
                        WHEN 'Staged' THEN 'Published'
                        ELSE 'MarkedForDeletion'
                    END,
                    update_timestamp = (CURRENT_TIMESTAMP AT TIME ZONE 'UTC'),
                    publish_timestamp = (CURRENT_TIMESTAMP AT TIME ZONE 'UTC')
                FROM input_splits
                WHERE
                    splits.index_uid = $1
                    AND splits.split_id = input_splits.split_id
                    AND NOT EXISTS (
                        SELECT 1
                        FROM input_splits
                        WHERE
                            actual_split_state != expected_split_state
                    )
            )
            -- Report the outcome of the update query.
            SELECT
                COUNT(1) FILTER (WHERE actual_split_state = 'Staged' AND expected_split_state = 'Staged'),
                COUNT(1) FILTER (WHERE actual_split_state = 'Published' AND expected_split_state = 'Published'),
                COALESCE(ARRAY_AGG(split_id) FILTER (WHERE actual_split_state IS NULL), ARRAY[]::TEXT[]),
                COALESCE(ARRAY_AGG(split_id) FILTER (WHERE actual_split_state != 'Staged' AND expected_split_state = 'Staged'), ARRAY[]::TEXT[]),
                COALESCE(ARRAY_AGG(split_id) FILTER (WHERE actual_split_state != 'Published' AND expected_split_state = 'Published'), ARRAY[]::TEXT[])
                FROM input_splits
        "#;
            let (
                num_published_splits,
                num_marked_splits,
                not_found_split_ids,
                not_staged_split_ids,
                not_marked_split_ids,
            ): (i64, i64, Vec<String>, Vec<String>, Vec<String>) =
                sqlx::query_as(PUBLISH_SPLITS_QUERY)
                    .bind(index_uid.to_string())
                    .bind(index_metadata_json)
                    .bind(staged_split_ids)
                    .bind(replaced_split_ids)
                    .fetch_one(tx)
                    .await
                    .map_err(|error| convert_sqlx_err(index_uid.index_id(), error))?;

            if !not_found_split_ids.is_empty() {
                return Err(MetastoreError::SplitsDoNotExist {
                    split_ids: not_found_split_ids,
                });
            }
            if !not_staged_split_ids.is_empty() {
                return Err(MetastoreError::SplitsNotStaged {
                    split_ids: not_staged_split_ids,
                });
            }
            if !not_marked_split_ids.is_empty() {
                return Err(MetastoreError::SplitsNotDeletable {
                    split_ids: not_marked_split_ids,
                });
            }
```
Expected split status is Published, but after first merge publish operation, the split status is already changed to MarkedDelete status.

**Comment by fulmicoton** [comment]
@geek-frio thank you for the great bug report!

I will come up with more questions soon, but can you confirm:
a) can you confirm the two merges happened on the same node?
b) can you share a larger slab of your logs (ideally containing the start of the merge operation for instance, was the timeline restarted, etc.)?

**Comment by geek-frio** [comment]
@fulmicoton the log on the indexer node and metastore node of that timeline is a little big. I'm not sure if you need it， here is the log in case you need to use them.
[ff6ca98f-90e5-44d9-bd28-d48a4e37f52a.csv](https://github.com/quickwit-oss/quickwit/files/12036136/ff6ca98f-90e5-44d9-bd28-d48a4e37f52a.csv)

**Comment by fulmicoton** [comment]
I just lied. Apparently we do not use the inventory for this.

**Comment by geek-frio** [comment]
> Thank you for the brilliant bug report. I haven't found the issue yet.
> 
> Flushing observations:
> 
> According to the logs it is just as @geek-frio described: business as usual, yet we schedule two merge targetting the same split.
> 
> ```
> 00:20:16.623Z,INFO,"quickwit_indexing::actors::merge_planner: Planned merge operation. merge_operation=Merge(operation_type=Merge, merged_split_id=01H569XD3FXSAREEERDGMR3ENA,splits=[01H569TYT79V4P5PNMTW2RK3JD,01H569TF9AETY17YZ23H3F4WP3,01H569SZV3BAXK0442YJJ8W3A0,])"
> 00:20:16.623Z,INFO,"quickwit_indexing::actors::merge_planner: Planned merge operation. merge_operation=Merge(operation_type=Merge, merged_split_id=01H569XD3FFVZAN807JSYJ710S,splits=[01H569VFSJZ9D7FT2NPV7J6JYZ,01H569V5P2PJH4CXBJKA84JWKJ,01H569VJCD3SCJP1RJV3D6P6Z6,])"
> 00:20:16.623Z,INFO,"merge{merge_split_id=01H569XD3FXSAREEERDGMR3ENA split_ids=[""01H569TYT79V4P5PNMTW2RK3JD"", ""01H569TF9AETY17YZ23H3F4WP3"", ""01H569SZV3BAXK0442YJJ8W3A0""] typ=Merge}:merge_split_downloader: quickwit_indexing::actors::merge_split_downloader: download-merge-splits dir=/quickwit/qwdata/indexing/loggerp-index%01H3RX5CM48GP9H0MYJBMWDSGK%loggerp-kafka-source%0%PVnkGC/merge%DrFpH0"
> 00:20:16.625Z,INFO,"quickwit_indexing::actors::merge_planner: Planned merge operation. merge_operation=Merge(operation_type=Merge, merged_split_id=01H569XD3GNY42PEFYKK6BX6XY,splits=[01H569V86PKGNJP961SHBTZ8T9,01H569TJG83JDT1JD9QB3GSBDE,01H569VJCEWZE39CY7QADWSE31,])"
> 
> 0:20:16.932Z,INFO,"merge{merge_split_id=01H569XD3FFVZAN807JSYJ710S split_ids=[""01H569VFSJZ9D7FT2NPV7J6JYZ"", ""01H569V5P2PJH4CXBJKA84JWKJ"", ""01H569VJCD3SCJP1RJV3D6P6Z6""] typ=Merge}:merge_split_downloader: quickwit_indexing::actors::merge_split_downloader: download-merge-splits dir=/quickwit/qwdata/indexing/loggerp-index%01H3RX5CM48GP9H0MYJBMWDSGK%loggerp-kafka-source%0%PVnkGC/merge%hqtddT"
> 00:20:16.933Z,INFO,"quickwit_indexing::actors::merge_planner: Planned merge operation. merge_operation=Merge(operation_type=Merge, merged_split_id=01H569XDD51DBRXZDC2NQCNZN1,splits=[01H569TF8F3R1W5W5G5WKS2Q5X,01H569SZW77VC3A60PBHC1RVGJ,01H569T354ZHR9T0JDSWSQMN3W,])"
> 
> 
> 00:20:16.945Z,INFO,"merge{merge_split_id=01H569XD3GNY42PEFYKK6BX6XY split_ids=[""01H569V86PKGNJP961SHBTZ8T9"", ""01H569TJG83JDT1JD9QB3GSBDE"", ""01H569VJCEWZE39CY7QADWSE31""] typ=Merge}:merge_split_downloader: quickwit_indexing::actors::merge_split_downloader: download-merge-splits dir=/quickwit/qwdata/indexing/loggerp-index%01H3RX5CM48GP9H0MYJBMWDSGK%loggerp-kafka-source%0%PVnkGC/merge%trcH2A"
> 00:20:16.946Z,INFO,"quickwit_indexing::actors::merge_planner: Planned merge operation. merge_operation=Merge(operation_type=Merge, merged_split_id=01H569XDDJ4Q0W1GVB6075DQ6S,splits=[01H569VX14MT4CHM013BHHM52E,01H569VE54SFPQBA7XCVN2SERV,01H569TYTK30VTP1JMH96E80YK,])"
> 
> 00:20:17.025Z,INFO,"merge{merge_split_id=01H569XD3FXSAREEERDGMR3ENA split_ids=[""01H569TYT79V4P5PNMTW2RK3JD"", ""01H569TF9AETY17YZ23H3F4WP3"", ""01H569SZV3BAXK0442YJJ8W3A0""] typ=Merge}:merge_executor: quickwit_indexing::actors::merge_executor: merge-operation-success merged_num_docs=18463 elapsed_secs=0.0932426 operation_type=Merge"
> 00:20:17.025Z,INFO,"merge{merge_split_id=01H569XD3FXSAREEERDGMR3ENA split_ids=[""01H569TYT79V4P5PNMTW2RK3JD"", ""01H569TF9AETY17YZ23H3F4WP3"", ""01H569SZV3BAXK0442YJJ8W3A0""] typ=Merge}:packager: quickwit_indexing::actors::packager: start-packaging-splits split_ids=[""01H569XD3FXSAREEERDGMR3ENA""]"
> 00:20:17.026Z,INFO,"merge{merge_split_id=01H569XDD51DBRXZDC2NQCNZN1 split_ids=[""01H569TF8F3R1W5W5G5WKS2Q5X"", ""01H569SZW77VC3A60PBHC1RVGJ"", ""01H569T354ZHR9T0JDSWSQMN3W""] typ=Merge}:merge_split_downloader: quickwit_indexing::actors::merge_split_downloader: download-merge-splits dir=/quickwit/qwdata/indexing/loggerp-index%01H3RX5CM48GP9H0MYJBMWDSGK%loggerp-kafka-source%0%PVnkGC/merge%baiaKZ"
> 00:20:17.026Z,INFO,"merge{merge_split_id=01H569XD3FXSAREEERDGMR3ENA split_ids=[""01H569TYT79V4P5PNMTW2RK3JD"", ""01H569TF9AETY17YZ23H3F4WP3"", ""01H569SZV3BAXK0442YJJ8W3A0""] typ=Merge}:packager: quickwit_indexing::actors::packager: create-packaged-split split_id=""01H569XD3FXSAREEERDGMR3ENA"""
> 
> 00:20:17.027Z,INFO,"quickwit_indexing::actors::merge_planner: Planned merge operation. merge_operation=Merge(operation_type=Merge, merged_split_id=01H569XDG31W7T691DTJKHW7WT,splits=[01H569WD97J2R1W33K5PEBH673,01H569W3M24ZPP18F1H5KVG43A,01H569VFSJZ9D7FT2NPV7J6JYZ,])"
> ```
> 
> The merge planner is supposed to be a single actor with view in RAM with the list of splits. It initiates itself from the Metastore, and then it gets its updates from the Publisher actors. Since it is an actor, its logic is single threaded. To avoid scheduling two merge on the same split, it has an Inventory object relying on Arc/Weak to get a perfect view of the splits that are locked. I suspect we have a race condition and drop this lock a bit too early.
> 
> According to the logs, the merge polanner was happy to recommend the merge after packaging, and before uploading the first merged split.

@fulmicoton Is it possible that the sending of messages here, due to a full queue, causes a loop break, and then in the subsequent code logic, PlanMerge is resent again, leading to multiple merges of the splits?

![图片](https://github.com/quickwit-oss/quickwit/assets/40055215/dfdedd69-c2ae-4a2e-b77c-de74d56cf458)

**Comment by fulmicoton** [comment]
> @fulmicoton Is it possible that the sending of messages here, due to a full queue, causes a loop break, and then in the subsequent code logic, PlanMerge is resent again, leading to multiple merges of the splits?

No, this cannot happen. Besides, the two merge operation that we are seeing in your logs have only one split in common, so this would not explain your problem.

A few extra questions:
- are you using a regular PostgreSQL or is it a Postgres compatible engine?
- can you confirm that qucikwit recovered from the error and continued indexing as usual?
- Would it be possible to have 5 more minutes of history of metastore + indexer logs?

**Comment by fmassot** [comment]
Additional context to @fulmicoton's questions: we saw this error `Splits `[""01H566NDGVQ2C945MFYQ8SXG3C""]` do not exist.)"` in the logs and we suspect we can dig more into this issue by understanding what happened before.

**Comment by geek-frio** [comment]
> > @fulmicoton Is it possible that the sending of messages here, due to a full queue, causes a loop break, and then in the subsequent code logic, PlanMerge is resent again, leading to multiple merges of the splits?
> 
> No, this cannot happen. Besides, the two merge operation that we are seeing in your logs have only one split in common, so this would not explain your problem.
> 
> A few extra questions:
> 
>     * are you using a regular PostgreSQL or is it a Postgres compatible engine?
> 
>     * can you confirm that qucikwit recovered from the error and continued indexing as usual?
> 
>     * Would it be possible to have 5 more minutes of history of metastore + indexer logs?


@fulmicoton @fmassot 
Sorry for the delay in responding and thanks for your help.
- We are using a normal postgresql database in Aliyun,the database version is :Postgresql 15.0, the image below is the description of the corresponding database information.
![图片](https://github.com/quickwit-oss/quickwit/assets/40055215/3ae28648-e04c-498d-8611-9f09cef69603)
- Yes, the quickwit will recover from this error and continue indexing.In our production environment, we encounter this error several times a day, and each time the issue resolves itself automatically.
- Below file is the log from an earlier time.
[463c846a-552b-4b54-b5ef-c47ca8dc50de.csv.zip](https://github.com/quickwit-oss/quickwit/files/12091598/463c846a-552b-4b54-b5ef-c47ca8dc50de.csv.zip)

**Comment by fulmicoton** [comment]
Thank you for the quick answer!

**Comment by fulmicoton** [comment]
@geek-frio  It has been fixed in #3854 thanks to @dojiong
